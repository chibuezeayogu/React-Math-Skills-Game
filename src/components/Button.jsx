import React, { Component } from 'react';
import PropTypes from 'prop-types';

/**
 * 
 * @class Button
 * 
 * @extends Component
 */
class Button extends Component {
  /**
   * 
   * @method handelAcceptAnswer
   * 
   * @memberof Button
   * 
   * @returns {void}
   */
  handleAcceptAnswer = () => {
    this.props.acceptAnswer();
  }

  /**
   * 
   * @method handelCheckAnswer
   * 
   * @memberof Button
   * 
   * @returns {void}
   */
  handleCheckAnswer = () => {
    this.props.checkAnswer();
  }

  /**
   * 
   * @method handelRedraw
   * 
   * @memberof Button
   * 
   * @returns {void}
   */
  handleRedraw = () => {
    this.props.redraw();
  }

  render() {
    const { selectedNumbers, redraws, answerIsCorrect } = this.props;
    let button;
    switch (answerIsCorrect) {
      case true:
        button = <button 
                  className="btn btn-outline-success" 
                  onClick={this.handleAcceptAnswer}>
                  <i className="fa fa-check"></i>
                </button>;
        break;
      case false:
        button = <button className="btn btn-outline-danger">
                  <i className="fa fa-times"></i>
                </button>;
        break;
      default:
        button = <button className="btn btn-outline-secondary"
                  onClick={this.handleCheckAnswer}
                  disabled={selectedNumbers.length === 0}>
                  =
                </button>;
    }
  
    return (
      <div>
        {button}
        <br /><br />
        <button 
          className="btn btn-outline-warning"
          type="button"
          onClick={this.handleRedraw}
          disabled={redraws === 0}
        >
          <i className="fas fa-redo-alt" aria-hidden="true"></i> {redraws}
        </button>
      </div>
    );
  }
}

Button.propTypes = {
  acceptAnswer: PropTypes.func,
  checkAnswer: PropTypes.func,
  redraw: PropTypes.func,
  redraws: PropTypes.number,
  selectedNumbers: PropTypes.array,
  answerIsCorrect: PropTypes.oneOfType([
    PropTypes.object,
    PropTypes.bool
  ]),
};

export default Button;
