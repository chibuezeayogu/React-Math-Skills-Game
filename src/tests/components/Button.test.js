/* 
global jest 
expect 
*/
import React from 'react';
import { shallow } from 'enzyme';

import Button from '../../components/Button';

describe('Button Component', () => {
  const setup = (status) => {
    const props = {
      answerIsCorrect: status,
      checkAnswer: jest.fn(),
      acceptAnswer: jest.fn(),
      selectedNumbers: [1],
      redraw: jest.fn(),
      redraws: 5
    };

    return shallow(<Button {...props}/>);
  };

  
  it('render correctly', () => {
    const wrapper = setup(null);
    expect(wrapper).toMatchSnapshot();
  });

  it('should call handelCheckAnswer when answerIsCorrect is ture', () => {
    const wrapper = setup(true);
    const handleSpy = jest.spyOn(wrapper.instance(), 'handelCheckAnswer');
    wrapper.instance().handelCheckAnswer();
    expect(handleSpy).toHaveBeenCalled();
  });

  it('should call handelCheckAnswer when answerIsCorrect is false', () => {
    const wrapper = setup(false);
    const handleSpy = jest.spyOn(wrapper.instance(), 'handelCheckAnswer');
    wrapper.instance().handelCheckAnswer();
    expect(handleSpy).toHaveBeenCalled();
  });

  it('should call handelAcceptAnswer method', () => {
    const wrapper = setup(true);
    const handleSpy = jest.spyOn(wrapper.instance(), 'handelAcceptAnswer');
    wrapper.instance().handelAcceptAnswer();
    expect(handleSpy).toHaveBeenCalled();
  });

  it('should call handelRedraw method', () => {
    const wrapper = setup(null);
    const handleSpy = jest.spyOn(wrapper.instance(), 'handelRedraw');
    wrapper.instance().handelRedraw();
    expect(handleSpy).toHaveBeenCalled();
  });

});
