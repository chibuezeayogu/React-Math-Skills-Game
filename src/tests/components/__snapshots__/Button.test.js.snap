// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component render correctly 1`] = `
<div
  className="col-2"
>
  <button
    className="btn btn-outline-secondary"
    disabled={false}
    onClick={[Function]}
  >
    =
  </button>
  <br />
  <br />
  <button
    className="btn btn-outline-warning"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <i
      className="fa fa-refresh"
    />
     
    5
  </button>
</div>
`;
